node/collation-generation/src/tests.rs

```diff
@@ -14,472 +14,1263 @@
 // You should have received a copy of the GNU General Public License
 // along with Polkadot.  If not, see <http://www.gnu.org/licenses/>.
 
-mod handle_new_activations {
-	use super::super::*;
-	use ::test_helpers::{dummy_hash, dummy_head_data, dummy_validator};
-	use futures::{
-		lock::Mutex,
-		task::{Context as FuturesContext, Poll},
-		Future,
-	};
-	use polkadot_node_primitives::{
-		BlockData, Collation, CollationResult, MaybeCompressedPoV, PoV,
-	};
-	use polkadot_node_subsystem::{
-		errors::RuntimeApiError,
-		messages::{AllMessages, RuntimeApiMessage, RuntimeApiRequest},
-	};
-	use polkadot_node_subsystem_test_helpers::{
-		subsystem_test_harness, TestSubsystemContextHandle,
-	};
-	use polkadot_primitives::{
-		CollatorPair, Id as ParaId, PersistedValidationData, ScheduledCore, ValidationCode,
+use super::*;
+use assert_matches::assert_matches;
+use futures::{
+	lock::Mutex,
+	task::{Context as FuturesContext, Poll},
+	Future,
+};
+use polkadot_node_primitives::{BlockData, Collation, CollationResult, MaybeCompressedPoV, PoV};
+use polkadot_node_subsystem::{
+	errors::RuntimeApiError,
+	messages::{AllMessages, RuntimeApiMessage, RuntimeApiRequest},
+};
+use polkadot_node_subsystem_test_helpers::{subsystem_test_harness, TestSubsystemContextHandle};
+use polkadot_node_subsystem_util::TimeoutExt;
+use polkadot_primitives::{
+	vstaging::{BackingState, CandidatePendingAvailability, Constraints, InboundHrmpLimitations},
+	BlockNumber, CandidateHash, CollatorPair, Id as ParaId, PersistedValidationData, ScheduledCore,
+	ValidationCode,
+};
+use sp_keyring::sr25519::Keyring as Sr25519Keyring;
+use std::pin::Pin;
+use test_helpers::{dummy_hash, dummy_head_data, dummy_validator};
+
+type VirtualOverseer = TestSubsystemContextHandle<CollationGenerationMessage>;
+
+fn test_harness<T: Future<Output = VirtualOverseer>>(test: impl FnOnce(VirtualOverseer) -> T) {
+	let pool = sp_core::testing::TaskExecutor::new();
+	let (context, virtual_overseer) =
+		polkadot_node_subsystem_test_helpers::make_subsystem_context(pool.clone());
+	let subsystem = async move {
+		let subsystem = crate::CollationGenerationSubsystem::new(Metrics::default());
+
+		subsystem.run(context).await;
 	};
-	use std::pin::Pin;
-
-	fn test_collation() -> Collation {
-		Collation {
-			upward_messages: Default::default(),
-			horizontal_messages: Default::default(),
-			new_validation_code: None,
-			head_data: dummy_head_data(),
-			proof_of_validity: MaybeCompressedPoV::Raw(PoV { block_data: BlockData(Vec::new()) }),
-			processed_downward_messages: 0_u32,
-			hrmp_watermark: 0_u32.into(),
-		}
+
+	let test_fut = test(virtual_overseer);
+
+	futures::pin_mut!(test_fut);
+	futures::pin_mut!(subsystem);
```

Comment:
Why do we pin `subsystem`?

I see that `test_fut` gets moved into an async block. So maybe that's why it needs to be pinned.

But `subsystem` doesn't appear to be explicitly moved.

Reply:
There's some trait barriers within `join` that seem to require this, AFAIK

Reply:
(I think there used to be, as this code was copied from elsewhere, but looks like they've relaxed that now)